articles/how-to-setup-user-authentication-from-scratch-with-rails-6/index.md

```diff
@@ -0,0 +1,702 @@
+```rb
+class WelcomeController < ApplicationController
+  def index
+  end
+end
+```
+
+The generator also created a file in `app/views/welcome/index.html.erb`, by default Rails renders a view that corresponds to the name of a controller action.
+Replace the view content with:
+
+```erb
+<h1>Rails Authentication</h1>
+```
+
+- We have our controller and views in place, let's create the model.
+
+A `model` is a `Ruby` class that hold data, we define it using a generator as follows:
+
+```bash
+rails generate model User email:string password_digest:string
+```
+
+This will generate several files, we will only focus on..`db/migrate/<timestamp>_create_users.rb` and the `app/models/users.rb`
+
+- Update `db/migrate/<timestamp>_create_users.rb` to:
+
+```rb
+class CreateUsers < ActiveRecord::Migration[6.1]
+  def change
+    create_table :users do |t|
+    t.string :email, null: false
+    t.string :password_digest
+
+    t.timestamps
+  end
+end
+```
+
+We are adding model level validation to our `email` field.
+`password_digest` is used to create password fields in rails and encryption is done by the [bcyrpt_gem](https://rubygems.org/gems/bcrypt). Include the gem in your Gemfile.
+
+- Install by running `bundle install`
+
+- Update `app/models/users.rb` to:
+
+```rb
+class User < ApplicationRecord
+   # adds virtual attributes for authentication
+   has_secure_password
+
+   # validates email
+   validates :email, presence: true, uniqueness: true, format: { with: /\A[^@\s]+@[^@\s]+\z/, message: 'Invalid email' }
+end
+```
+
+- Remember to run your migrations.
+
+```bash
+rails db:migrate
+```
+
+Test our app,
+
+```bash
+rails server
+```
+
+We will see a Rails app with a home page.
+
+![localhost](/engineering-education/how-to-setup-user-authentication-from-scratch-with-rails-6/localhost.png)
+
+Test our model.
+
+```bash
+rails console
+```
+
+Create a new user
+
+![railsconsole](/engineering-education/how-to-setup-user-authentication-from-scratch-with-rails-6/railsconsole.png)
+
+### Configuring routes
+
+- Update `config/routes.rb`
+
+```rb
+Rails.application.routes.draw do
+  root 'welcome#index'
+
+  get 'sign_up', to: 'registrations#new'
+  post 'sign_up', to: 'registrations#create'
+
+  get 'sign_in', to: 'sessions#new'
+  post 'sign_in', to: 'sessions#create', as: 'log_in'
+
+  delete 'logout', to: 'sessions#destroy'
+
+  get 'password', to: 'passwords#edit', as: 'edit_password'
+  patch 'password', to: 'passwords#update'
+
+  get 'password/reset', to: 'password_resets#new'
+  post 'password/reset', to: 'password_resets#create'
+  get 'password/reset/edit', to: 'password_resets#edit'
+  patch 'password/reset/edit', to: 'password_resets#update'
+end
+```
+
+We are matching our routes to the above controller and actions
+
+### Adding Controllers
+
+Routers determine which controller to use for the request,controllers will then receive the request and save or fetch data from our models.
+
+- touch `app/controllers/registrations_controller.rb`
```

Comment:
Say, Let's create the` registrations_controller.rb` by running the command `touch app/controllers/registrations_controller.rb`

Reply:
Make sure you do this for the other controllers too.